Prevent races resulting in incorrect search index results for newly sent messages.

pratik-pc opened the pull request:
This PR checks message id against pending message ids in `fts_update_log` and calculates narrow match directly if it is yet to be indexed in the db. This prevents false negative narrow match on newly sent message which sometimes results in not adding to the current narrow even if it should.

Fixes: zulip#33267

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
\* **Fts indexing is disabled here**
| Before | After |
|-----------|---------|
| <video src="https://github.com/user-attachments/assets/9245eb68-ceb1-4fa3-9457-bc7af3003d7a"> | <video src="https://github.com/user-attachments/assets/e9699b5b-3a67-46b1-b376-4e1082e8f1a2">

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

pratik-pc committed:
migrations: Add message_id index in fts_update_log.

This commit adds a message_id index on fts_update_log to make it
faster to search pending fts indexing message ids in fts_update_log.

timabbott commented on zerver/migrations/0740_index_fts_message_id.py:
```diff
@@ -0,0 +1,20 @@
+                DROP INDEX CONCURRENTLY IF EXISTS fts_update_log_message_id
+            """,
+        ),
+    ]
```
Missing newline at the end of the file.

timabbott commented on zerver/views/message_fetch.py:
```diff
@@ -369,8 +374,18 @@ def messages_in_narrow_backend(
+                .select_from(table("fts_update_log"))
+                .where(column("message_id", Integer).in_(msg_ids))
+            )
+            pending_result = sa_conn.execute(pending_query)
```
I guess the benefit of this, if we return a `CodedError` for this case, is that it's a fast, just-index lookup, and doesn't do extra work.

pratik-pc committed:
narrow: Calculate search match directly in db query for unindexed fts.

Earlier, when narrowing to a search and a message event was added to
the `fts_update_log` for fts indexing, it would sometimes give false
negative for message ids which are yet to be indexed. This would end
up not showing the message in the current search narrow even if it
matches.

This commit fixes this behaviour by directly calculating if it matches
the search term during database query.

Fixes: zulip#33267.

opmkumar commented:
Looks good to me in manual testing.
@zulipbot remove "buddy review"

pratik-pc commented:
@ligmitz This is pretty small, only to make sure the sql is correct

Hell1213 commented:
hey @pratik-pc are you still working on this issue ???

pratik-pc commented:
> I have a question @pratik-pc , are we looking to check if the message is present in the indexed database and if not then find in fts_update_log, right? Is this an exclusive OR condition or a normal OR condition ?

Just to answer this, it is using normal OR condition but essentially an "exclusive" OR operation

alya commented:
@mateuszmandera Tagging you for review on this one.

mateuszmandera commented:
@pratik-pc Sorry for losing track of this one! Can you rebase this on main and then I can review?
